src/core/thread/network_data_leader_ftd.cpp

```diff
@@ -1319,9 +1319,45 @@ void Leader::HandleNetworkDataRestoredAfterReset(void)
 {
     const PrefixTlv *prefix;
     TlvIterator      tlvIterator(GetTlvsStart(), GetTlvsEnd());
+    ChangedFlags     flags;
 
     mWaitingForNetDataSync = false;
 
+    // Remove entries in Network Data from any un-allocated Router ID.
+    // This acts as a safeguard against an edge case where the leader
+    // is reset at an inopportune time, such as right after it removed
+    // an allocated router ID and sent MLE advertisement but before it
+    // got the chance to send the updated Network Data to other
+    // routers.
+
+    while (true)
+    {
+        bool     didChange = false;
+        Iterator iterator  = kIteratorInit;
+        uint16_t rloc16;
+
+        while (GetNextServer(iterator, rloc16) == kErrorNone)
+        {
+            if (!Get<RouterTable>().IsAllocated(Mle::RouterIdFromRloc16(rloc16)))
+            {
+                // After we `RemoveRloc()` the `iterator` is no longer valid
+                // so we `break` from the inner loop so to go over entries
+                // again from the outer `while (true)` loop.
+
+                RemoveRloc(rloc16, kMatchModeRouterId, flags);
+                didChange = true;
```

Comment:
Can we simplify the nested while loops by setting `iterator` back to `kIteratorInit` here?

Reply:
We can but I guess we would then be indirectly relying on behavior of `GetNextServer()` when the `iterator` when it returns an `error` (when net data is empty or we reach end of all entries). I think this is the case in current implementation of `GetNextServer()`, but I feel may be better not to assume this behavior when we use it here.

Reply:
Don't we already rely on `GetNextServer()` to return error to break out of the loop? Can you provide an example on some alternative behavior that you have in mind?

Reply:
@jwhui I think I misunderstood your suggestion. 😄... Yes. I think what you suggest would work nicely.

I assumed you want us to use  `if (iterator == kInteratorInit)` as way to check if we exited early (replacing `didChange` check)

Reply:
Changed it in new push (now we only have one loop). Thanks 😃